### ISSUE TITLE
Won't start on PocketBook 613

### ISSUE THREAD

**Issue by TenMirim** [koreader/koreader#4029]
Firstly, i know the app isn't designed for this low-end reader, just wanted to warn others that it **won't** work there since it crashes on start (from what i read from crash report it seems it can't detect it). 
* KOReader version: beta 20180602
* Device: PocketBoot 613 Basic
It doesn't even load up, stops on the clock icon, crashes and returns to the previous menu.
Crash log: [crash.log](https://github.com/koreader/koreader/files/2150056/crash.log)

**Comment by Frenzie** [comment]
To test if it works you'll just need to add the codename. For example, change `elseif codename == "PB627" then` to `elseif codename == "PocketBook 613" then`.

https://github.com/koreader/koreader/blob/ca6c7316e6fa550f511db72f9c1cdefd1775a2fe/frontend/device/pocketbook/device.lua#L295-L296

**Comment by TenMirim** [comment]
Oh, looks like it booted after creating PB 613 device in `frontend/PocketBook/device.lua` , but the app can't be controlled, because it probably uses different/non-standard key layout. Turning the "has-touch" option off didn't help. Just for context, it has only D-pad, button in middle of the D-pad, power button and two page navigating buttons, but can't find any official names for them.

**Comment by Frenzie** [comment]
If you add `-d` to the script and touch all the keys it'll log their values.

You did put in hasKeys = yes? And it actually doesn't have a touch screen?

NB There's also something about screen refreshes, so please grab the latest nightly instead of the stable.

**Comment by TenMirim** [comment]
No, it really doesn't have a touch screen. And it looks like it's receiving something, i can turn pages, but can't access menu or anything, really. Where am I supposed to put the argument? Somewhere in the .app file?

**Comment by Frenzie** [comment]
Indeed, if you add -d on this line like so:

https://github.com/koreader/koreader/blob/8ecdf9d88814e19ba6b3aeae2db9167a70f90740/platform/pocketbook/koreader.app#L47

```
./reader.lua -d "${args}" >>crash.log 2>&1 
```

The key codes seem to be defined here:
https://github.com/koreader/koreader/blob/8ecdf9d88814e19ba6b3aeae2db9167a70f90740/frontend/device/pocketbook/device.lua#L21-L38

You seem to be having an issue with the menu key?

**Comment by TenMirim** [comment]
From observation it looks like the MENU button is assigned as event of type 25/26 and code 10 (i pressed the button the most and it's the most frequent event), the left Dpad is 19,up Dpad is 17, right Dpad is 20, and down Dpad 18, quite straightforward.
Now, is here any way to assign buttons per device?

**Comment by Frenzie** [comment]
25 and 26 are quite expected. ;-)

https://github.com/koreader/koreader/blob/8ecdf9d88814e19ba6b3aeae2db9167a70f90740/frontend/device/pocketbook/device.lua#L13-L16

Code 10 is currently mapped to OK:
https://github.com/koreader/koreader/blob/8ecdf9d88814e19ba6b3aeae2db9167a70f90740/frontend/device/pocketbook/device.lua#L23

It looks like there are only three keys currently assigned to actions:
https://github.com/koreader/koreader/blob/8ecdf9d88814e19ba6b3aeae2db9167a70f90740/frontend/device/pocketbook/device.lua#L66-L70

So for testing, just change that to something like:

```lua
        event_map = {
            [KEY_OK] = "Menu",
            [KEY_MENU] = "Menu",
            [KEY_PREV] = "LPgBack",
            [KEY_NEXT] = "LPgFwd",
},
```

However, on further reflection that's one button too few to actually be properly useful I think. You said there's only the four arrows, an ok button, and pg up/down, correct? In that case I suppose mapping `[KEY_OK] = "Press",` might do the most.

The problem is that at the minimum you need four arrows and pg up/down to properly navigate the program, a click/press/enter button, and also a menu button. (Actually you might still get stuck then without an escape/back button.)

Perhaps if you remap left to `Menu` and up to `Back` (or vice versa) you might have a chance. So that'd make for something like:

```lua
        event_map = {
            [KEY_UP] = "Menu",
            [KEY_LEFT] = "Back",
            [KEY_OK] = "Press",
            [KEY_MENU] = "Menu",
            [KEY_PREV] = "LPgBack",
            [KEY_NEXT] = "LPgFwd",
},
```

That way you can't navigate up or left in dialogs, but I think most of them loop around and those that don't should be closable with "Back".

**Comment by TenMirim** [comment]
And isn't here anything like LT_PG overlapping in function with D-pad L? That way i might bind stuff on D-pad L/R and still navigate properly. Ik that exactly allows the firmware to do in standard menus.

**Comment by Frenzie** [comment]
What is LT_PG overlapping?

**Comment by TenMirim** [comment]
like the turn page left and left on the Dpad for example.

**Comment by Frenzie** [comment]
You mean with one of them acting as a modifier, like Shift on a keyboard? Except somehow not being a modifier at the same time? (Introducing a slight delay?)

**Comment by TenMirim** [comment]
like just plainly being able to use Turn to left and Turn to Right keys in menus, and sign the menu button to those two (or any other btns needed) and we might get away with not having enough buttons.

**Comment by Frenzie** [comment]
The main problem with that is that these keys are already bound to actions. In the file browser, right brings up the context menu, as distinct from page down which switches to the next page. It's not just a simple global change either because we wouldn't want to diminish the experience on a Kindle NT4/keyboard/gamepad.

Maybe @onde2rock can think of something. See https://github.com/koreader/koreader/issues/3818#issue-310217534 regarding the Kindle NT4 button layout.

**Comment by TenMirim** [comment]
And is here anything like long press detection created like it was mentioned in the issue referenced? That way you can in theory double the action modes of buttons, aka for example using right for next page and right hold for context menu? I know that wouldn't pass globally, but is it possible to setup something of this custom per device? Since the code looks quite hectic to me I don't know if and how to accomplish creating custom scheme for certain device.

**Comment by onde2rock** [comment]
Yes, two more keys on your device would be ideal (menu and back).

I already looked in long press. It's way more complicated since it should be done in C, and change the current behavior (from button down => immediate action ). And it feel laggy.

However, since Next_Page and Prev_Page are almost never used in the menu, maybe they could be remap to Back and Menu, when in menu.

And an unused arrow key to open the menu.

You should be able to get at least the bare minimum to use Koreader.

I will look into it.

**Comment by Frenzie** [comment]
That issue isn't about long-press detection on keyboard, but about the fact that the menu long-press action (or context action if you prefer) isn't currently accessible using keys.

Long press on keyboard would be absolutely horrible.

**Comment by onde2rock** [comment]
Agreed, long press on keyboard is hopeless.

I'm going to add some conditon in the code for this particular device, would you prefer something like 
`if Device:isPocketBook613`
or more general like 
`if Device:hasNotEnoughtKeys`

?

**Comment by Frenzie** [comment]
I think something more generic like `hasFewKeys` would be better in case there are more like that. (Perhaps a PocketBook sister/cousin model.)

**Comment by Frenzie** [comment]
@lenormf

> Or come up with an alternative PocketBook structure that implements those by default?

`hasFewKeys` is required to implement the appropriate behavior throughout the UI. At the same time you will still need device-specific stuff. There's no "or". ^_^

**Comment by lenormf** [comment]
Now you've lost me.

Will KOReader be made usable on devices that only have a DPad, somehow? I.e. without binding direction keys to unrelated actions like "Menu" or "Back"?

**Comment by Frenzie** [comment]
Nothing will just happen by itself of course. ;-) If you want to write stuff like long press and/or double tap that's fine by me; that's not contradictory to what @onde2rock sketched.

In a situation like this one I'd do it incrementally. I'd start very simple, like what I wrote in https://github.com/koreader/koreader/issues/4029#issuecomment-469330608. That's the exact same thing as what @poire-z stated in https://github.com/koreader/koreader/issues/6168#issuecomment-633247874, great minds and all that. ^_^ The key concept is not that it would provide a particularly good experience, but that it's the difference between completely unusable and some basic use hopefully being viable. Also because it's a very simple change to test out.

Then I'd expand that to what @onde2rock wrote in https://github.com/koreader/koreader/issues/4029#issuecomment-469387377. Similar concept, just much smarter.

If more nuance is still desired after those comparatively easy tweaks, you could work on long press and/or double tap. I'd be somewhat wary of it though, responding on keyup rather than keydown makes things feel slower. But it could be a necessary compromise.

Of course you don't have to follow the same path I would. The advantage to what I sketched is that you'd have something that's working along every step of the way and the mechanisms put in place in the earlier steps are pretty much required scaffolding for the follow-up steps regardless. I see it all as essentially complementary.

**Comment by rjd22** [comment]
A lot of work for devices like these has been done? Can someone confirm if the situation of non-touch devices has improved?

**Comment by Frenzie** [comment]
You probably simply didn't quite write what you mean, but note that devices with too few keys are something very different than non-touch. Non-touch works quite well, though you might run into an oversight if you go into some rarely accessed dialog, or a new one where we didn't pay proper attention to it. Most of that should be automatically taken care of by FocusManager. Few keys by contrast doesn't work particularly well, or in any case I don't think there's been much of anything done in the matter other than my extremely basic support.

Anyway, you can easily test by telling your emulator it is or isn't touch, does or doesn't have a d-pad, and has few keys. Navigate with arrow keys. Left functions as back. See #6195 for details.

The big problem with `Device:hasFewKeys()` is that it doesn't have a back button. Realistically the minimum to be properly usable with our (or frankly any) UI is something like back, menu, arrows, enter.

But strictly speaking you only need 2 arrow keys to reach everywhere rather inconveniently, hence using left as back. And generic non-touch improvements to FocusManager, as well as the likes of https://github.com/koreader/koreader/pull/9041 and https://github.com/koreader/koreader/pull/8712 also make more of the program accessible to `hasFewKeys()`.

**Comment by rjd22** [comment]
Ah thank you for your clarification. So the issue, if I understand correctly, is that there aren't enough hardware keys to navigate KoReader with the device properly and to do so alternate keybinds would have to introduced like long press or double press.

**Comment by Frenzie** [comment]
I implemented what I did because I believe that's vastly preferable over making everything feel sluggish, but long or double would be an alternative that could also perform the same task, yes.
